doc/man/Qremote.8
.\"TOPICS "Topics:"
.TH Qremote 8 "April 2016" "Qremote Version @QSMTP_VERSION@" "Qremote"
.SH NAME
Qremote \- send mail via (E)SMTP
.SH SYNOPSIS
.B Qremote
.I host
.I sender
.I recip
[
.I recip ...
]
.SH DESCRIPTION
.B Qremote
reads a mail message from its standard input and sends 
the message to one or more recipients at a remote host.
The standard input must be a descriptor of a file opened 
for reading, Qremote will fstat(2) and mmap(2) it.

The remote host is
.BR Qremote 's
first argument,
.IR host .
.B Qremote
sends the message to
.IR host ,
using the routing algorithm described below or to a mail exchanger for
.I host
listed in the Domain Name System, via the Simple Mail Transfer
Protocol (SMTP), where
.I host
can be either a fully-qualified domain name:

.EX
     mail.example.com
.EE

or an IP address enclosed in brackets:

.EX
     [192.0.2.19]
.EE

or an IPv6 address enclosed in brackets:

.EX
     [::ffff:192.0.2.19]
.EE

The envelope recipient addresses are listed as
.I recip
arguments to
.BR Qremote .
The envelope sender address is listed as
.I sender\fP.

.SH TRANSPARENCY
End-of-file in SMTP is encoded as dot CR LF.
A dot at the beginning of a line is encoded as dot dot.
It is impossible in SMTP to send a message that does not end with a newline.
.B Qremote
converts single CR or LF into the SMTP newline convention by converting them
to CRLF.

If the message contains non-ascii characters the message is encoded
as quoted-printable if the remote server does not support the 8BITMIME
extension defined in RfC 1652. If a message contains lines with more than
1000 characters it is always recoded to quoted-printable: sending such a
long line would be a SMTP protocol violation.

.B Qremote
supports the CHUNKING extension defined in RfC 3030. If the remote server
supports CHUNKING no message recoding is done at all.
.SH "RESULTS"
.B Qremote
prints some number of 
.I recipient reports\fP
followed by a
.I message report\fR.
Each report is terminated by a 0-byte.
Each report begins with a single letter:
.TP 5
r
Recipient report: acceptance.
.TP 5
h
Recipient report: permanent rejection.
.TP 5
s
Recipient report: temporary rejection.
.TP 5
K
Message report: success.
.I host
has taken responsibility for delivering the message to each
acceptable recipient.
.TP 5
Z
Message report: temporary failure.
.TP 5
D
Message report: permanent failure.
.PP
After this letter comes a human-readable description of
what happened. No human readable description is printed
after 'r' status report
.RB ( qmail-rspawn
would get confused otherwise).

The recipient reports will always be printed in the same order as
.BR Qremote 's
.I recip
arguments.
Note that in failure cases there may be fewer
recipient reports
than
.I recip
arguments.

.B Qremote
always exits zero.
.SH "CONTROL FILES"
The files listed in this section are looked up in the subdirectory
.I control/
of the qmail directory. All files are optional.

.TP 5
.I clientcert.pem
SSL certificate that is used to authenticate with the remote server
during a TLS session.
.TP 5

.I helohost
Current host name, for use solely in saying hello to the remote SMTP server.
Default:
.IR me ,
if that is supplied;
otherwise
.B Qremote
refuses to run.
.TP 5

.I outgoingip
The local IPv4 address used when sending out mails. Use this if your machine
has multiple addresses and you want your mail coming from a specific one.
.TP 5

.I outgoingip6
Works the same as
.IR outgoingip ,
but for IPv6 addresses.
.TP 5

.I smtproutes
Artificial SMTP routes. See below for a detailed description of mail routing.

The format of this file is the same as used by
.B qmail-remote .

Each route has the form
.IR hostpattern\fB:\fIrelay\fB[:\fIport\fB]\fI ,
without any extra spaces. The
.I port
specification is optional. The following 2 lines have the same effect:

.EX
   foo.example.com:mx2.example.com
   foo.example.com:mx2.example.com:25
.EE

The first matching entry will be used.

.I hostpattern
may start with a dot, matching all subdomains of the given domain,
but not domain itself.

.I relay
may be empty, causing the default routing mechanisms to be applied,
this can be used to either only specify a different destination port
or to route one domain using the default mechanism that would otherwise
be matched by a later rule. The following example will route
foo.example.com using the default mechanism, but any other subdomain of
example.com to mail.example.net. example.com is also routed using the
default mechanism as it is not matched by any rule:

.EX
   foo.example.com:
   .example.com:mail.example.net
.EE

.TP 5
.I smtproutes.d

.RS
Artificial SMTP routes. This is a more extensible format than the plain
.I smtproutes
file, allowing easier addition of other properties for the target host
while keeping the files readable. See below for a detailed description of
mail routing.

A file can be created for each domain to specify a
specific routing. Wildcard routings may be created using
a file beginning with a single
.I *
as subdomain, e. g.

.EX
  *.example.net
.EE

Make sure to escape the * character when accessing these files from a
shell as the shell will use this character as glob extension.

Each file contains
.I key=value
pairs defining the mapping. Currently the following keys are supported:

.RS
.IP relay 0.8i
The hostname of the target machine.
.IP port
The port on the remote machine to connect to.
.IP clientcert
The TLS certificate to use when connecting.
.IP clientkey
The TLS key to use when connecting. If not given
.I clientcert
is used.
.IP outgoingip
The IPv4 address used when connecting to the remote server.
.IP outgoingip6
The IPv6 address used when connecting to the remote server.
.RE

Duplicate keys are forbidden. Lines beginning with a hash sign '#' and empty
lines are ignored.

Given the domain
.I foo.example.net
the following lookups will be performed:

.EX
 smtproutes.d/foo.example.net
 smtproutes.d/*.example.net
 smtproutes.d/*.net
 smtproutes.d/default
.EE

The search is terminated on the first match.
Please note that a wildcard-subdomain does not match messages
with the absolute destination of the remainder, e. g.
.I *.example.net
does not match a message sent to
.I foo@example.net
.

Only values from one file are considered; if a value is not set in the first
matching file no further files will be scanned for that value. Every value that
is not explicitly given will be taken from the default control files as
specified above.

The following example would route mails to port 24 on
.I baz.example.org
:

.EX
   host=baz.example.org
   port=24
.EE
.RE

.TP 5
.I timeoutconnect
Number of seconds
.B Qremote
will wait for the remote SMTP server to accept a connection.
Default: 60.
The kernel normally imposes a 75-second upper limit.
.TP 5
.I timeoutremote
Number of seconds
.B Qremote
will wait for each response from the remote SMTP server.
Default: 1200.

.TP 5
.I tlsclientciphers
A set of OpenSSL client cipher strings. Multiple ciphers
contained in a string should be separated by a colon.

.TP 5
.I tlshosts/<FQDN>.pem
.B Qremote
requires authentication from servers for which this certificate exists
.RB ( <FQDN>
is the fully-qualified domain name of the server). One of the
.I DNSname
or the
.I CommonName
attributes have to match.

.B WARNING:
this option may cause mail to be delayed, bounced, doublebounced, or lost.

.SH "MAIL ROUTING"

.RS
By default
.B Qremote
will look up the host given as it's first argument in the DNS system, looking
for MX entries. If none such entry is found it will look for AAAA or A records
for this hostname and use them. If neither is found, a temporary error will
be announced, causing the mail to remain in the queue or eventually being
bounced if it has stayed there for too long.

The routing may be influenced using the
.I smtproutes
control file or the files in the
.I smtproutes.d
directory. See above for a description of the syntax rules that apply to these
files.

.RS
.nr step 1 1
The routing algorithm is as follows:
.IP \n[step] 4
search for a matching entry in
.I smtproutes.d
.IP \n+[step]
search for a matching entry in
.I smtproutes
.IP \n+[step]
search for MX entries for the target host
.IP \n+[step]
search for AAAA and A entries for the target host
.RE

If a match in the first 2 steps is found the
.I relay
specified
.B Qremote
will connect to
.IR relay ,
as if
.I host
had
.I relay
as its only MX.
(It will also avoid doing any CNAME lookups on
.IR recip .)

If
.I relay
is empty this tells
.B Qremote
to look up DNS records as usual, but apply any other properties
specified in the control files (e.g. different target port).

The
.B qmail
system does not protect you if you create an artificial
mail loop between machines.
However,
you are always safe using
.I smtproutes
if you do not accept mail from the network.
.RE

.SH DEBUGGING
If
.B Qremote
is compiled with the
.I DEBUG_IO
flag the contents of the SMTP transmissions can be recorded. They will
be sent to the syslog daemon with facility mail and log level debug. The contents of the SMTP DATA phase will
never be logged for privacy reasons.

Logging is not enabled by default. If the file
.I control/Qremote_debug
is readable on startup logging will be enabled. Therefore it will usually not harm to
compile that facility into the program.

.SH "SEE ALSO"
fstat(2),
mmap(2),
addresses(5),
envelopes(5),
qmail-control(5),
qmail-send(8),
Qsmtpd(8),
qmail-tcpok(8),
qmail-tcpto(8)
.SH AUTHOR
Rolf Eike Beer
.SH WEBSITE
https://opensource.sf-tec.de/Qsmtp/


qremote/qremote.c
/** \file qremote.c
 \brief main functions of Qremote

 This file contains the main function, the configuration and error handling of Qremote,
 the drop-in replacement for qmail-remote.
 */

#include <qremote/qremote.h>

#include <control.h>
#include <ipme.h>
#include <log.h>
#include <netio.h>
#include <qdns.h>
#include <qmaildir.h>
#include <qremote/conn.h>
#include <qremote/greeting.h>
#include <qremote/qrdata.h>
#include <qremote/starttlsr.h>
#include <sstring.h>
#include <tls.h>

#include <errno.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <signal.h>
#include <stdio.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <syslog.h>
#include <unistd.h>

int socketd = -1;
unsigned int smtpext;	/**< the SMTP extensions supported by the remote server */
char *rhost;		/**< the DNS name (if present) and IP address of the remote server to be used in log messages */
size_t rhostlen;	/**< valid length of rhost */
char *partner_fqdn;	/**< the DNS name of the remote server (forward-lookup), or NULL if the connection was done by IP */

/**
 * @brief send QUIT to the remote server and close the connection
 *
 * This will properly shut down the connection to the remote server but will
 * not terminate the program.
 */
void
quitmsg(void)
{
	netwrite("QUIT\r\n");
	do {
		/* don't care about what he replies: we want to quit, if he don't want us to he must pay money *eg* */
		if (net_read(0)) {
			log_write(LOG_ERR, "network read error while waiting for QUIT reply");
			break;
		}
	} while ((linein.len >= 4) && (linein.s[3] == '-'));
	if (ssl) {
		ssl_free(ssl);
		ssl = NULL;
	}
	close(socketd);
	socketd = -1;

	free(partner_fqdn);
	partner_fqdn = NULL;
	free(rhost);
	rhost = NULL;
	free_smtproute_vals();
}

void
net_conn_shutdown(const enum conn_shutdown_type sd_type)
{
	if ((sd_type == shutdown_clean) && (socketd >= 0)) {
		quitmsg();
	} else if (socketd >= 0) {
		close(socketd);
		socketd = -1;

		if (ssl != NULL) {
			ssl_free(ssl);
			ssl = NULL;
		}

		free(partner_fqdn);
		free(rhost);
		free_smtproute_vals();
	}

#ifdef USESYSLOG
	closelog();
#endif

	free(heloname.s);
	if (msgdata != MAP_FAILED)
		munmap((void*)msgdata, msgsize);

	exit(0);
}

void
err_mem(const int doquit)
{
	write_status("Z4.3.0 Out of memory.");

	net_conn_shutdown(doquit ? shutdown_clean : shutdown_abort);
}

void
err_conf(const char *errmsg)
{
	const char *msg[] = {errmsg, NULL};
	err_confn(msg, NULL);
}

/**
 * @brief log a configuration error and exit
 * @param errmsg array of strings to combine to the message to log
 * @param freebuf a pointer to a buffer passed to free() after logging
 *
 * Use freebuf if the contents of this buffer need to be part of errmsg.
 * If you do not have anything to free just pass NULL.
 */
void
err_confn(const char **errmsg, void *freebuf)
{
	log_writen(LOG_ERR, errmsg);
	free(freebuf);

	write_status("Z4.3.0 Configuration error.");
	net_conn_shutdown(shutdown_clean);
}

static void
setup(void)
{
#ifdef USESYSLOG
	openlog("Qremote", LOG_PID, LOG_MAIL);
#endif

	/* Block SIGPIPE, otherwise the process will get killed when the remote
	 * end cancels the connection improperly. */
	sigset_t mask;
	sigemptyset(&mask);
	sigaddset(&mask, SIGPIPE);

	if (sigprocmask(SIG_BLOCK, &mask, NULL) == -1) {
		write_status("Z4.3.0 Cannot block SIGPIPE.");

		net_conn_shutdown(shutdown_abort);
	}

	remote_common_setup();

#ifdef CHUNKING
	unsigned long chunk;
	if (loadintfd(openat(controldir_fd, "chunksizeremote", O_RDONLY | O_CLOEXEC), &chunk, 32768) < 0) {
		err_conf("parse error in control/chunksizeremote");
	} else {
		if (chunk >= ((unsigned long)1 << 31)) {
			err_conf("chunksize in control/chunksizeremote too big");
		}
		chunksize = chunk & 0xffffffff;
	}
#endif

#ifdef DEBUG_IO
	do_debug_io = (faccessat(controldir_fd, "Qremote_debug", R_OK, 0) == 0);
#endif
}

int
main(int argc, char *argv[])
{
	struct ips *mx = NULL;
	int rcptcount = argc - 3;
	struct stat st;

	/* do this check before opening any files to catch the case that fd 0 is closed at this point */
	int i = fstat(0, &st);

	setup();

	if (rcptcount <= 0) {
		log_write(LOG_CRIT, "too few arguments");
		write_status("Z4.3.0 internal error: Qremote called with invalid arguments");
		net_conn_shutdown(shutdown_abort);
	}

	/* this shouldn't fail normally: qmail-rspawn did it before successfully */
	if (i != 0) {
		if (errno == ENOMEM)
			err_mem(0);
		log_write(LOG_CRIT, "can't fstat() input");
		write_status("Z4.3.0 internal error: can't fstat() input");
		net_conn_shutdown(shutdown_abort);
	}
	msgsize = st.st_size;
	msgdata = mmap(NULL, msgsize, PROT_READ, MAP_SHARED, 0, 0);

	if (msgdata == MAP_FAILED) {
		log_write(LOG_CRIT, "can't mmap() input");
		write_status("Z4.3.0 internal error: can't mmap() input");
		net_conn_shutdown(shutdown_abort);
	}

	getmxlist(argv[1], &mx);
	if (targetport == 25) {
		mx = filter_my_ips(mx);
		if (mx == NULL) {
			const char *msg[] = { "Z4.4.3 all mail exchangers for ",
					argv[1], " point back to me" };
			write_status_m(msg, 3);
			net_conn_shutdown(shutdown_abort);
		}
	}
	sortmx(&mx);

	i = connect_mx(mx, &outgoingip, &outgoingip6);
	freeips(mx);

	if (i < 0) {
		write_status("Z4.4.2 can't connect to any server");
		net_conn_shutdown(shutdown_abort);
	}

	if (ssl) {
		successmsg[3] = "message ";
		successmsg[4] = SSL_get_cipher(ssl);
		if (SSL_get0_dane_tlsa(ssl, NULL, NULL, NULL, NULL, NULL) > 0)
			successmsg[5] = " encrypted and DANE secured";
		else
			successmsg[5] = " encrypted";
	}

/* check if message is plain ASCII or not */
	const unsigned int recodeflag = need_recode(msgdata, msgsize);

	if (send_envelope(recodeflag, argv[2], argc - 3, argv + 3) != 0)
		net_conn_shutdown(shutdown_clean);

	successmsg[0] = rhost;
#ifdef CHUNKING
	if (smtpext & esmtp_chunking) {
		send_bdat(recodeflag);
	} else {
#else
	{
#endif
		send_data(recodeflag);
	}
	net_conn_shutdown(shutdown_clean);
}
